oe_content.module

```diff
@@ -106,3 +109,128 @@ function oe_content_entity_base_field_info(EntityTypeInterface $entity_type) {
+    '#options' => [
+      1 => t('Enabled'),
+      0 => t('Disabled'),
+    ],
+  ];
+
+  $form['#entity_builders'][] = 'oe_content_form_field_config_form_builder';
+
+}
+
+/**
+ * Entity form builder for the field_config form.
+ *
+ * Saves the composite field settings into the field config settings.
+ *
+ * @param string $entity_type
+ *   The name of the entity type.
+ * @param \Drupal\field\FieldConfigInterface $field_config
+ *   The field config.
+ * @param array $form
+ *   The form array.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The form state.
+ */
+function oe_content_form_field_config_form_builder(string $entity_type, FieldConfigInterface $field_config, array &$form, FormStateInterface $form_state): void {
+  if ($form_state->getValue('composite')) {
+    $field_config->setThirdPartySetting('oe_content', 'composite', TRUE);
+  }
+  else {
+    $field_config->setThirdPartySetting('oe_content', 'composite', FALSE);
+  }
+}
+
+/**
+ * Implements hook_entity_delete().
+ *
+ * Delete any composite referenced entities when deleting an entity.
+ */
+function oe_content_entity_delete(EntityInterface $entity): void {
+  if (!$entity instanceof \Drupal\Core\Entity\FieldableEntityInterface) {
+    return;
+  }
+  /** @var \Drupal\Core\Field\FieldDefinitionInterface[] $field_definitions */
+  $field_definitions = \Drupal::service('entity_field.manager')->getFieldDefinitions($entity->getEntityTypeId(), $entity->bundle());
+  $applicable_field_types = ['entity_reference', 'entity_reference_revisions'];
+  foreach ($field_definitions as $key => $definition) {
+    if (!in_array($definition->getType(), $applicable_field_types)) {
+      continue;
+    }
+    if ($definition instanceof FieldConfigInterface && $definition->getThirdPartySetting('oe_content', 'composite', FALSE)) {
+      $referenced_entities = $entity->get($key)->referencedEntities();
+      /** @var \Drupal\Core\Entity\EntityInterface $referenced_entity */
+      foreach ($referenced_entities as $referenced_entity) {
+        if ($referenced_entity->id() !== $entity->id() && oe_content_referenced_entity_deletable($referenced_entity)) {
+          $referenced_entity->delete();
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Checks whether we can delete a referenced entity or not.
+ *
+ * @param \Drupal\Core\Entity\EntityInterface $entity
+ *   The entity being checked.
+ *
+ * @return bool
+ *   TRUE if the entity is not referneced anywhere else, FALSE otherwise.
+ */
+function oe_content_referenced_entity_deletable(EntityInterface $entity): bool {
+  $entity_type_manager = \Drupal::entityTypeManager();
+
+  // Getting fields which could have references to the entity
+  // we are trying to delete.
+  $fields = $entity_type_manager->getStorage('field_config')->loadByProperties([
+    'field_type' => 'entity_reference',
+  ]);
+  $fields = array_merge($fields, $entity_type_manager->getStorage('field_config')->loadByProperties([
```

Comment:
lol, why don't you just loadByProperties(['field_type', ['entity_reference', 'entity_reference_revisions']])?

Reply:
I tried that but loading field types by properties doesn't use the IN sql selector, so it ends up returning nothing....